### PR TITLE
Cudagraph always on and cudagraphs fallback flags

### PR DESCRIPTION
**Pull request by voznesenskym**

### CODE REVIEW

File: torchinductor/config.py [hunk 1]
```diff
@@ -49,6 +49,12 @@
 # replace small reductions with pointwise, disable with `= 1`
 unroll_reductions_threshold = 8
 
+# Cudagraphify will run even if we have graph mutations
+always_attempt_cudagraphify = True
```
Comment (anijain2305) [inline torchinductor/config.py:53]: Just for my education? Why have this flag? Is this to prefetch other problems?

Comment (voznesenskym) [reply torchinductor/config.py:53]: There are cases in prod models where we over-enthusiastically decide that a graph was mutated and then abort attempting cudagraphify. This just lets it always have a go.

**Comment by voznesenskym** [comment]
This may not work - if it causes a segfault - still iterating, but up for feedback

File: torchinductor/config.py [hunk 2]
```diff
@@ -49,6 +49,13 @@
 # replace small reductions with pointwise, disable with `= 1`
 unroll_reductions_threshold = 8
 
+# Cudagraphify will run even if we have graph mutations, may segfault
+always_attempt_cudagraphify = False
+
+# Cudagraphify failures will not stop compilation, but will not produce cudagraphified code
+# Will not proceed past segfaults
+fallback_cudagraphify = True
```
Comment (voznesenskym) [inline torchinductor/config.py:57]: Maybe default False for debugging?

File: torchinductor/config.py [hunk 3]
```diff
@@ -49,6 +49,13 @@
 # replace small reductions with pointwise, disable with `= 1`
 unroll_reductions_threshold = 8
 
+# Cudagraphify will run even if we have graph mutations, may segfault
```
Comment (jansel) [inline torchinductor/config.py:52]: I believe this will be silently incorrect, not segfault.

When would the user want this behavior?

Comment (voznesenskym) [reply torchinductor/config.py:52]: @anijain2305 asked the same thing above.

There are cases in prod models where we think we are over-enthusiastically deciding that a graph was mutated and then abort attempting cudagraphify. This just lets it always have a go. 

However, we are using fail/no fail as the indicator of success - not any correctness testing. Is there a way to check that what was produced is correct?

Comment (jansel) [reply torchinductor/config.py:52]: The graph will return the correct answer, but the mutations to the inputs will be lost.

I might call this `force_cudagraphs` or something similar -- then update the comment to say it discards input mutations.

Comment (voznesenskym) [reply torchinductor/config.py:52]: Sounds good, it should feel relatively unsafe to use, and force fits the idea.

File: torchinductor/compile_fx.py [hunk 4]
```diff
@@ -60,15 +60,28 @@ def compile_fx_inner(
...
-    ):
-        compiled_fn = cudagraphify(
-            compiled_fn, example_inputs, static_input_idxs=range(num_fixed)
-        )
+    )
+
+    # TODO(voz): Add a config driven option to install a segfault handler for the sake of reporting.
+    run_cudagraphs = safe_cudagraphs or config.force_cudagraphs
+    if config.force_cudagraphs and not safe_cudagraphs:
+        log.warning("Cudagraph failure heuristics bypassed, may segfault.")
```
Comment (jansel) [inline torchinductor/compile_fx.py:72]:
```suggestion
        log.warning("Cudagraph failure heuristics bypassed, may segfault or not perform input mutation")
```

File: torchinductor/compile_fx.py [hunk 5]
```diff
@@ -60,15 +60,28 @@ def compile_fx_inner(
...
+    if cudagraphs and run_cudagraphs:
+        try:
+            cudagraph_fn = cudagraphify(
+                compiled_fn, example_inputs, static_input_idxs=range(num_fixed)
+            )
+        except Exception as e:
+            if config.fallback_cudagraphify:
+                log.warning("Cudagraph failed, but running with fallback enabled.")
```
Comment (jansel) [inline torchinductor/compile_fx.py:80]: If we continue we need to disable cudagraphs in the backwards with `BoxedBool.disable(cudagraphs)`

File: torchinductor/compile_fx.py [hunk 6]
```diff
@@ -60,15 +60,28 @@ def compile_fx_inner(
...
+            else:
+                raise e
+
+        compiled_fn = cudagraph_fn
```
Comment (jansel) [inline torchinductor/compile_fx.py:84]: I don't think this is needed, we can just set it directly on L75

**Comment by jansel** [comment]
We have migrated `torchdynamo` to `torch._dynamo` and will use the `pytorch/pytorch` repo for future development.  Please resubmit this PR to https://github.com/pytorch/pytorch/

More details and instructions to port this PR over can be found in https://github.com/pytorch/torchdynamo/issues/1588

**Pull request closed without merging**